### PR TITLE
Using declaration ioperation

### PR DESCRIPTION
**Pull request by chsienki**
IOperation and CFG support for using declarations

### CODE REVIEW

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for review please

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 1]
```diff
@@ -2899,4 +2899,21 @@
       </Comments>
     </Property>
   </Node>
+  <Node Name="IUsingVariableDeclarationOperation" Base="IOperation">
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: >IUsingVariableDeclarationOperation [](start = 14, length = 34)

It looks like syntactically a Using Declaration is just a Local Declaration with additional  keywords. Should we simply reuse the existing node, IVariableDeclarationOperation, which already says that it is used to represent "(3) C# using declarations"?

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: `IVariableDeclarationGroupOperation` is used as a child node of the regular using statement. So you have a parent `IUsingOperation` with a VariableDeclarationGroup and a body. That's the same for the other types mentioned in the comments.

This way we keep the same shape for using declarations: you have a parent using declaration, which contains a child declaration group of the variables declared. 

I originally went the route of updating DeclarationGroup with an enum on it to specify if it was using/async using/none, but this proved simpler as it was never used in any of the other cases, and this way you can keep the tree shape the same and visit the using declaration essentially the same as the statement.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: I am not following why would we need to make any changes to DeclarationGroup. Or why similarities with IUsingOperation are important. I think we should simply add IsUsing and IsAsynchronous properties to IVariableDeclarationOperation instead of adding new node. 

---
In reply to: [332695447](https://github.com/dotnet/roslyn/pull/39125#discussion_r332695447) [](ancestors = 332695447)

File: src/Compilers/Core/Portable/Generated/Operations.Generated.cs [hunk 2]
```diff
@@ -2767,6 +2768,24 @@ internal interface IWithOperation : IOperation
...
         IOperation Value { get; }
     }
+    /// <summary>
+    /// Represents a using variable declaration statement.
+    /// </summary>
+    /// <remarks>
+    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
+    /// change it in the future.
+    /// </remarks>
+    public interface IUsingVariableDeclarationOperation : IOperation
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Generated/Operations.Generated.cs]: Dumb question.  Why is this a new iop, as opposed to being a bit on IVariableDeclarationGroupOperation ?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -1364,28 +1364,28 @@ private IOperation FinishVisitingStatement(IOperation originalOperation, IOperat
             return result ?? MakeInvalidOperation(originalOperation.Syntax, originalOperation.Type, ImmutableArray<IOperation>.Empty);
         }
 
-        private void VisitStatements(ArrayBuilder<IOperation> statements)
+        private void VisitStatements(IList<IOperation> statements)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: why this change.  wont' this cause interface dispatches now?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: I agree with @AlekseyTs here. That's how i would have expected thsi would be represented.

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: Oh I see what you're referring to, that was from the initial 'dummy' implementation we did. Yep, we could definitely do that if you think it's better than introducing a new node?

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: I agree with reusing the `IVariableDeclarationOperation` for this, though I'm not sure I agree with the two bool flags. These aren't independent variables: if `IsAsync` were true, `IsUsing` _must_, by definition, be true. I would likely prefer an enum over two bools. This will also give us flexibility if we add new types of variable declarations in the future, as we can add new kinds rather than throwing a whole bunch of flag booleans on the type itself.

**Comment by 333fred** [comment]
As noted offline, we need tests for async using declarations.

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml (hunk 1 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: Yeah, enum is how I originally designed it interestingly enough, then split it out into its own operation, because it felt weird having most of the declarations have `UsingType=None`

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: I wouldn't call it `UsingType`. I'd call it it `DeclarationType`, with options of `Standard`, `Using`, and `AsyncUsing`, or something along those lines.

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 4]
```diff
@@ -2036,6 +2035,13 @@
         </summary>
       </Comments>
     </Property>
+    <Property Name="UsingKind" Type="UsingKind">
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: I don't think this is what I'd do either. `UsingKind` limits future extensibility in the same was that the two flag bools do. Rather, I'd suggest something like `DeclarationKind`, with options of `Local`, `Using`, and `AsyncUsing`.

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: For example, there's an open request for a `fixed` declaration statement, that would fit in nicely with this API.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: Any reason these aren't just bools?  `IsUsing` `IsAsync`?  (and, in the future `IsFixed`)?

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: `DeclarationKind` is problematic because it already exists as public in Microsoft.CodeAnalysis.Editing  
`LocalDeclarationKind` already exists in Microsoft.CodeAnalysis.VisualBasic.Symbols although thats `Friend` so we could conceivable promote it and re-use it for this purpose?

We could do the two bools solution, but is a bit odd as you can't have IsAsync be true without IsUsing.

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: @333fred i only know DeclarationKind already exists because that's how I first defined it. Non clashing name suggestions welcome, and i'll happily update

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: @CyrusNajmabadi https://github.com/dotnet/roslyn/pull/39125#discussion_r333158890.
I'll mull on the DeclarationKind naming.

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: Actually, I thought it collided a lot more than it did. I've changed it to `DeclarationKind`

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: I don't feel hugely strongly.  I'm personally ok with booleans where one implies the other.  But i'm not going to die on this hill :)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 3 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Derp. Fixed.

File: src/Compilers/Core/Portable/Generated/Operations.Generated.cs (hunk 2 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/Generated/Operations.Generated.cs]: Resolved

File: src/Compilers/Core/Portable/Operations/DeclarationKind.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Operations
+{
+    public enum DeclarationKind
+    {
+        /// <summary>
+        /// A local declaration.
+        /// </summary>
+        Local = 0,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/DeclarationKind.cs]: >Local [](start = 8, length = 5)

Are other kinds not local? Perhaps "Default" or "None"?

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 6]
```diff
@@ -475,6 +475,10 @@ public override void VisitVariableDeclaration(IVariableDeclarationOperation oper
         {
             var variableCount = operation.Declarators.Length;
             LogString($"{nameof(IVariableDeclarationOperation)} ({variableCount} declarators)");
+            if (operation.DeclarationKind != DeclarationKind.Local)
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: >if (operation.DeclarationKind != DeclarationKind.Local) [](start = 12, length = 55)

Please make necessary adjustments to TestOperationVisitor

File: src/Tools/Source/CompilerGeneratorTools/CompilerTools.sln [hunk 7]
```diff
@@ -1,17 +1,19 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
```
Comment (AlekseyTs) [inline src/Tools/Source/CompilerGeneratorTools/CompilerTools.sln]: I am assuming changes to this file were not intentional. Please revert/exclude from the PR.

File: src/Compilers/Core/Portable/Operations/DeclarationKind.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Operations
+{
+    public enum DeclarationKind
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/DeclarationKind.cs]: >DeclarationKind [](start = 16, length = 15)

VariableDeclarationKind?

File: src/Tools/Source/CompilerGeneratorTools/CompilerTools.sln (hunk 7 above)
Comment (chsienki) [reply src/Tools/Source/CompilerGeneratorTools/CompilerTools.sln]: There were intentional. The CompilerTools.sln didn't have the new CompilersIOperationGenerator added to it.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 9]
```diff
@@ -224,6 +224,7 @@ internal IOperation CreateInternal(BoundNode boundNode)
                 case BoundKind.LocalDeclaration:
                     return CreateBoundLocalDeclarationOperation((BoundLocalDeclaration)boundNode);
                 case BoundKind.MultipleLocalDeclarations:
+                case BoundKind.UsingLocalDeclarations:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >case BoundKind.UsingLocalDeclarations: [](start = 16, length = 38)

It looks like changes made to bound trees to support Using Declarations violate the invariant that only leaf nodes can have kinds. I.e. base nodes shouldn't have kinds. This way we can be sure that type checks for nodes that have kinds are equivalent to kind checks.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 10]
```diff
@@ -1755,16 +1754,19 @@ private IVariableDeclarationGroupOperation CreateBoundMultipleLocalDeclarationsO
             SyntaxNode declarationSyntax = declarationGroupSyntax.IsKind(SyntaxKind.LocalDeclarationStatement) ?
                     ((LocalDeclarationStatementSyntax)declarationGroupSyntax).Declaration :
                     declarationGroupSyntax;
+
+            bool isUsing = declarationGroupSyntax.IsKind(SyntaxKind.LocalDeclarationStatement) && ((LocalDeclarationStatementSyntax)declarationGroupSyntax).UsingKeyword != default;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >declarationGroupSyntax.IsKind(SyntaxKind.LocalDeclarationStatement) && ((LocalDeclarationStatementSyntax)declarationGroupSyntax).UsingKeyword != default; [](start = 27, length = 153)

We shouldn't be relying on syntax in IOperation factory. All required information should be captured in bound nodes during binding and factory should rely only on information in bound nodes.

**Comment by AlekseyTs** [comment]
>             if (boundNode is BoundUsingLocalDeclarations boundUsingLocalDeclarations)

Should this code be removed now?

---
Refers to: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:349 in 07733a2. [](commit_id = 07733a29ce7e2126423ccd14b2a8c1ae3488831c, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 11]
```diff
@@ -1755,16 +1754,19 @@ private IVariableDeclarationGroupOperation CreateBoundMultipleLocalDeclarationsO
...
 
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
             // If the syntax was the same, we're in a fixed statement or using statement. We make the Group operation implicit in this scenario, as the
             // syntax itself is a VariableDeclaration
-            // We do the same if the declarationSyntax was a using declaration, as it's bound as if it were a using statement
-            bool isUsing = declarationGroupSyntax.IsKind(SyntaxKind.LocalDeclarationStatement) && ((LocalDeclarationStatementSyntax)declarationGroupSyntax).UsingKeyword != default;
-            bool isImplicit = declarationGroupSyntax == declarationSyntax || boundMultipleLocalDeclarations.WasCompilerGenerated || isUsing;
+            bool isImplicit = declarationGroupSyntax == declarationSyntax || boundMultipleLocalDeclarations.WasCompilerGenerated;
             return new VariableDeclarationGroupOperation(ImmutableArray.Create(multiVariableDeclaration), _semanticModel, declarationGroupSyntax, type, constantValue, isImplicit);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >return new VariableDeclarationGroupOperation(ImmutableArray.Create(multiVariableDeclaration), _semanticModel, declarationGroupSyntax, type, constantValue, isImplicit); [](start = 12, length = 167)

I guess I was confused before about what node maps to the variable declaration statement. Since it looks like ```IVariableDeclarationGroupOperation``` maps to a statement, I think the new property should be added to that node instead.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 9 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: So do you want me to change this as part of this PR, or should we just log a bug and fix it in the future?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 10 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Agreed, but we were already relying on syntax here to distinguish between using/fixed and localdecl. Should we just create an issue to fix the underlying bound nodes (as well as splitting them out above) and link it here?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: > Should we just create an issue to fix the underlying bound nodes (as well as splitting them out above) and link it here?

I don't think so. Picking best syntax to associate node with is one thing, reinterring semantic meaning (which should be clear from the bound node itself) from syntax is another.

---
In reply to: [333679279](https://github.com/dotnet/roslyn/pull/39125#discussion_r333679279) [](ancestors = 333679279)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 9 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: > So do you want me to change this as part of this PR, or should we just log a bug and fix it in the future?

I think it is a simple fix - introduction of a common base node. Should fit into this PR.

---
In reply to: [333678166](https://github.com/dotnet/roslyn/pull/39125#discussion_r333678166) [](ancestors = 333678166)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 12]
```diff
@@ -1364,28 +1364,29 @@ private IOperation FinishVisitingStatement(IOperation originalOperation, IOperat
...
         private void VisitStatements(ImmutableArray<IOperation> statements)
         {
-            foreach (var statement in statements)
+            for (int i = 0; i < statements.Length; i++)
             {
-                VisitStatement(statement);
-            }
-        }
+                if (statements[i] is VariableDeclarationGroupOperation declarationGroup
+                    && declarationGroup.Declarations[0].DeclarationKind != DeclarationKind.Local)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >&& declarationGroup.Declarations[0].DeclarationKind != DeclarationKind.Local [](start = 20, length = 76)

It would be good to not make assumption that all ```Declarations``` have the same declaration kind. Moving the property to the group should take care of that.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 13]
```diff
@@ -6623,7 +6627,7 @@ public override IOperation VisitInvalid(IInvalidOperation operation, int? captur
             IOperation visitInvalidOperationStatement(IInvalidOperation invalidOperation)
             {
                 Debug.Assert(_currentStatement == invalidOperation);
-                VisitStatements(children);
+                VisitStatements(children.ToImmutable());
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >ToImmutable [](start = 41, length = 11)

ToImmutableAndFree and adjust other places to do the same?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 9 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Done.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 10 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 11 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Extracted to group instead

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 13 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Do we want to actually free the children though? They aren't owned by the walker so must be freed somewhere else?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 14]
```diff
@@ -6835,6 +6839,36 @@ public override IOperation VisitSwitchExpression(ISwitchExpressionOperation oper
...
         }
 
+        private void VisitUsingVariableDeclarationOperation(VariableDeclarationGroupOperation operation, ImmutableArray<IOperation> statements)
+        {
+            IOperation saveCurrentStatement = _currentStatement;
+            _currentStatement = operation;
+            StartVisitingStatement(operation);
+
+            var declarationKind = operation.DeclarationKind;
+            Debug.Assert(declarationKind != DeclarationKind.Default);
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider instead asserting that the kind is either `Using` or `AsynchronousUsing`

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 15]
```diff
@@ -6835,6 +6839,36 @@ public override IOperation VisitSwitchExpression(ISwitchExpressionOperation oper
...
+
+            // a using statement introduces a 'logical' block after declaration, we synthesize one here in order to analyze it like a regular using 
+            BlockOperation logicalBlock = new BlockOperation(
+                operations: statements,
+                locals: default,
+                operation.OwningSemanticModel,
+                operation.Syntax,
+                operation.Type,
+                operation.ConstantValue,
+                isImplicit: true);
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Nit: extra newline below.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 16]
```diff
@@ -2105,5 +2080,16 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
             bool isImplicit = placeholder.WasCompilerGenerated;
             return new InstanceReferenceOperation(referenceKind, _semanticModel, syntax, type, constantValue, isImplicit);
         }
+
+        private IVariableDeclarationGroupOperation CreateBoundUsingLocalDeclarationOperation(BoundUsingLocalDeclarations boundUsingLocalDeclarations)
+        {
+            bool isAsync = boundUsingLocalDeclarations.AwaitOpt is object;
+            DeclarationKind declarationKid = (isAsync) ? DeclarationKind.AsynchronousUsing : DeclarationKind.Using;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >(isAsync) [](start = 45, length = 9)

Nit: extraneous parens

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs [hunk 17]
```diff
@@ -2427,25 +2427,25 @@ void M1()
     IVariableDeclarationOperation (1 declarators) (OperationKind.VariableDeclaration, Type: null, IsInvalid) (Syntax: 'int[y switc ...  new int[0]')
       Ignored Dimensions(1):
           ISwitchExpressionOperation (1 arms) (OperationKind.SwitchExpression, Type: System.Int32, IsInvalid) (Syntax: 'y switch { int z => 42 }')
-            Value:
+            Value:
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs]: Nit: are the only changes in this file whitespace? Can they be reverted?

File: src/Compilers/Core/Portable/Operations/DeclarationKind.cs (hunk 8 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/DeclarationKind.cs]: I also like this suggestion, it follows the rest of our Kind enums.

---
In reply to: [333643366](https://github.com/dotnet/roslyn/pull/39125#discussion_r333643366) [](ancestors = 333643366)

**Comment by 333fred** [comment]
Done review pass (commit 14). Overall looks good, just a couple of small things to address.

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 18]
```diff
@@ -2002,7 +2009,6 @@
         Current Usage:
           (1) C# VariableDeclaration
           (2) C# fixed declarations
-          (3) C# using declarations
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: >(3) C# using declarations [](start = 10, length = 25)

Consider renumbering remaining items below

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 13 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Do we want to actually free the children though? They aren't owned by the walker so must be freed somewhere else?

We are in a local function and ```children``` is a builder created by the containing method. It is freed before we return from the method. It looks like instead of calling ```Free``` before the return, we could call ```ToImmutableAndFree``` here and in ```visitInvalidOperationExpression```, replacing ```invalidOperation.Children.ToImmutableArray()``` with ```children.ToImmutableAndFree()```.  

---
In reply to: [333732853](https://github.com/dotnet/roslyn/pull/39125#discussion_r333732853) [](ancestors = 333732853)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 19]
```diff
@@ -6835,6 +6839,35 @@ public override IOperation VisitSwitchExpression(ISwitchExpressionOperation oper
...
+            _currentStatement = operation;
+            StartVisitingStatement(operation);
+
+            var declarationKind = operation.DeclarationKind;
+            Debug.Assert(declarationKind == VariableDeclarationKind.Using || declarationKind == VariableDeclarationKind.AsynchronousUsing);
+
+            // a using statement introduces a 'logical' block after declaration, we synthesize one here in order to analyze it like a regular using 
+            BlockOperation logicalBlock = new BlockOperation(
+                operations: statements,
+                locals: default,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Do we ever create blocks with default locals? Perhaps using an empty array would be more robust.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 20]
```diff
@@ -6835,6 +6839,35 @@ public override IOperation VisitSwitchExpression(ISwitchExpressionOperation oper
...
+                operation.OwningSemanticModel,
+                operation.Syntax,
+                operation.Type,
+                operation.ConstantValue,
+                isImplicit: true);
+
+            HandleUsingOperationParts(
+                resources: operation,
+                body: logicalBlock,
+                locals: default,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >default [](start = 24, length = 7)

Same comment about using an empty array instead.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 21]
```diff
@@ -2105,5 +2080,16 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
             bool isImplicit = placeholder.WasCompilerGenerated;
             return new InstanceReferenceOperation(referenceKind, _semanticModel, syntax, type, constantValue, isImplicit);
         }
+
+        private IVariableDeclarationGroupOperation CreateBoundUsingLocalDeclarationOperation(BoundUsingLocalDeclarations boundUsingLocalDeclarations)
+        {
+            bool isAsync = boundUsingLocalDeclarations.AwaitOpt is object;
+            VariableDeclarationKind declarationKid = isAsync ? VariableDeclarationKind.AsynchronousUsing : VariableDeclarationKind.Using;
+            SyntaxNode declarationSyntax = ((LocalDeclarationStatementSyntax)boundUsingLocalDeclarations.Syntax).Declaration;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >(LocalDeclarationStatementSyntax) [](start = 44, length = 33)

It would be more robust to not assume that the cast is going to succeed. For example, a compiler generated node doesn't have to be associated on a local declaration statement syntax.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 22]
```diff
@@ -2105,5 +2080,16 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
...
+            bool declarationIsImplicit = boundUsingLocalDeclarations.WasCompilerGenerated;
+
+            IVariableDeclarationOperation multiVariableDeclaration = new CSharpLazyVariableDeclarationOperation(this, boundUsingLocalDeclarations, _semanticModel, declarationSyntax, type: null, constantValue: default, declarationIsImplicit);
+            return new VariableDeclarationGroupOperation(ImmutableArray.Create(multiVariableDeclaration), declarationKid, _semanticModel, boundUsingLocalDeclarations.Syntax, type: null, constantValue: default, isImplicit: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >false [](start = 222, length = 5)

It is not clear why we simply assume that the node is not implicit. 
In general, it looks like we can share code with ```CreateBoundMultipleLocalDeclarationsOperation```. That method can be changed to take BoundMultipleLocalDeclarationsBase and figure out declaration kind based on the actual kind of the node, plus accessing BoundUsingLocalDeclarations.AwaitOpt if necessary.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 16)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 23]
```diff
@@ -2105,5 +2086,16 @@ private IInstanceReferenceOperation CreateCollectionValuePlaceholderOperation(Bo
             bool isImplicit = placeholder.WasCompilerGenerated;
             return new InstanceReferenceOperation(referenceKind, _semanticModel, syntax, type, constantValue, isImplicit);
         }
+
+        //private IVariableDeclarationGroupOperation CreateBoundUsingLocalDeclarationOperation(BoundUsingLocalDeclarations boundUsingLocalDeclarations)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >//private IVariableDeclarationGroupOperation CreateBoundUsingLocalDeclarationOperation(BoundUsingLocalDeclarations boundUsingLocalDeclarations) [](start = 8, length = 143)

Please remove commented out code.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 17)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 24]
```diff
@@ -1364,28 +1364,29 @@ private IOperation FinishVisitingStatement(IOperation originalOperation, IOperat
...
+                        followingStatements.Add(statements[j]);
+                    }
 
-        private void VisitStatements(IEnumerable<IOperation> statements)
-        {
-            foreach (var statement in statements)
-            {
-                VisitStatement(statement);
+                    // visit the using decl with the following statements
+                    VisitUsingVariableDeclarationOperation(declarationGroup, followingStatements.ToImmutableAndFree());
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >followingStatements.ToImmutableAndFree() [](start = 77, length = 40)

Consider using `ImmutableArray.Create(statements, i + 1, statements.Length - i)`.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 25]
```diff
@@ -1743,29 +1734,35 @@ private IOperation CreateBoundLocalDeclarationOperation(BoundLocalDeclaration bo
...
             bool isImplicit = (varStatement == varDeclaration) || boundLocalDeclaration.WasCompilerGenerated;
-            return new VariableDeclarationGroupOperation(ImmutableArray.Create(multiVariableDeclaration), _semanticModel, varStatement, type, constantValue, isImplicit);
+            return new VariableDeclarationGroupOperation(ImmutableArray.Create(multiVariableDeclaration), VariableDeclarationKind.Default, _semanticModel, varStatement, type, constantValue, isImplicit);
         }
 
-        private IVariableDeclarationGroupOperation CreateBoundMultipleLocalDeclarationsOperation(BoundMultipleLocalDeclarations boundMultipleLocalDeclarations)
+        private IVariableDeclarationGroupOperation CreateBoundMultipleLocalDeclarationsBaseOperation(BoundMultipleLocalDeclarationsBase boundMultipleLocalDeclarations)
         {
-            // The syntax for the boundMultipleLocalDeclarations can either be a LocalDeclarationStatement or a VariableDeclaration, depending on the context
+            // The syntax for the boundMultipleLocalDeclarationsBase can either be a LocalDeclarationStatement or a VariableDeclaration, depending on the context
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >Base [](start = 64, length = 4)

The parameter name does not include "Base".

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 26]
```diff
@@ -1743,29 +1734,35 @@ private IOperation CreateBoundLocalDeclarationOperation(BoundLocalDeclaration bo
...
             // We generate a DeclarationGroup for these scenarios (using/fixed) to maintain tree shape consistency across IOperation.
             SyntaxNode declarationGroupSyntax = boundMultipleLocalDeclarations.Syntax;
             SyntaxNode declarationSyntax = declarationGroupSyntax.IsKind(SyntaxKind.LocalDeclarationStatement) ?
                     ((LocalDeclarationStatementSyntax)declarationGroupSyntax).Declaration :
                     declarationGroupSyntax;
+
             bool declarationIsImplicit = boundMultipleLocalDeclarations.WasCompilerGenerated;
             IVariableDeclarationOperation multiVariableDeclaration = new CSharpLazyVariableDeclarationOperation(this, boundMultipleLocalDeclarations, _semanticModel, declarationSyntax, null, default, declarationIsImplicit);
 
+            // If this is a using declaration, work out it's declaration kind
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >it's [](start = 56, length = 4)

Typo: "its". Or perhaps use "the".

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 27]
```diff
@@ -1364,28 +1364,24 @@ private IOperation FinishVisitingStatement(IOperation originalOperation, IOperat
...
         {
-            foreach (var statement in statements)
+            for (int i = 0; i < statements.Length; i++)
             {
-                VisitStatement(statement);
+                if (statements[i] is VariableDeclarationGroupOperation declarationGroup
+                    && declarationGroup.DeclarationKind != VariableDeclarationKind.Default)
+                {
+                    // visit the using decl with the following statements
+                    var followingStatements = ImmutableArray.Create(statements, Math.Min(i + 1, statements.Length - 1), statements.Length - i - 1);
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >Math.Min(i + 1, statements.Length - 1) [](start = 80, length = 38)

Won't this copy the last statement in the block, even if that is the `using` statement?

Comment (cston) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: `ImmutableArray.Create(statements, i + 1, statements.Length - i)` should work, even if `i + 1 == statements.Length`.

---
In reply to: [335196847](https://github.com/dotnet/roslyn/pull/39125#discussion_r335196847) [](ancestors = 335196847)

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: No, because the length becomes zero in that case and you end up with an empty array. I've added `UsingDeclaration_Flow_11` to test this scenario.

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: But what about when there is only one statement in the list, which is a using decl? `i = 0 => i + 1 = 1, statements.Length - i == 1` which is incorrect. 

---
In reply to: [335199463](https://github.com/dotnet/roslyn/pull/39125#discussion_r335199463) [](ancestors = 335199463,335196847)

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: No, because the length becomes zero in that case and you end up with an empty array. I've added `UsingDeclaration_Flow_11` to test this scenario.

---
In reply to: [335196847](https://github.com/dotnet/roslyn/pull/39125#discussion_r335196847) [](ancestors = 335196847)

Comment (chsienki) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Discussed offline + fixed

---
In reply to: [335205220](https://github.com/dotnet/roslyn/pull/39125#discussion_r335205220) [](ancestors = 335205220,335196847)
